Button Component: Update styling to avoid relying on Core Styles making it reusable

youknowriad opened the pull request:
Right now, the button component is relying heavily on core styles. This has some issues:

 - The need to add `.core-ui` to any place where we want to extend the default button styles
 - The components module is not reusable. One of the goals of the Gutenberg modules is to make them reusable in WP Admin and outside WP Admin as well. If we ever want to publish the components module on npm, it needs to define its own styles.
 - In components heavy application, we don't reuse classNames, we reuse components. 

This PR copies the styles from Core into the `components-button` className to solve this.

**Notes**

I probably need to update the default styles according to the WP theme cc @jasmussen (I might need some help here)

**Testing instructions**

Navigate in the Gutenberg UI and check that the styling of the different buttons didn't regress.

youknowriad committed:
Button Component: Update styling to avoid relying on Core Styles making it reusable

youknowriad commented on components/button/index.js:
```diff
@@ -47,12 +49,14 @@ class Button extends Component {
+			'is-small': isSmall,
 			'is-toggled': isToggled,
 			'is-busy': isBusy,
+			'is-link': isLink,
```
There are way too many button states here, but this matches Core, we can try to rethink some of them later.

noisysocks replied:
Would it make sense to replace all these booleans with a more enum-like API? e.g. `<Button style="link">`?

youknowriad replied:
I don't know maybe later, let's not change all the buttons today :)

afercia commented:
There's one thing I've noticed also previously and this PR improves it but still a bit weird:

when I try to make an IconButton have also some text, the position of the icon is OK if the button is a real button, but it's not OK when it's actually a link. Seems it happens in Chrome, Firefox is OK. Seems the `text-indent: 4px;` doesn't work for links in Chrome. See the Preview button I've made it an IconButton for testing purposes:

Chrome:

<img width="1050" alt="screen shot 2018-05-03 at 17 42 48" src="https://user-images.githubusercontent.com/1682452/39587821-c10f935e-4efa-11e8-81c0-47672118ba43.png">

Firefox:

<img width="1054" alt="screen shot 2018-05-03 at 17 42 51" src="https://user-images.githubusercontent.com/1682452/39587839-c9f596c6-4efa-11e8-903f-0dca11e74bab.png">

noisysocks reviewed (dismissed):
Agree 100% with the motivation behind this change. It seems to cause a few visual regressions, however.

Various icons are misaligned:

<img width="128" alt="screen shot 2018-05-04 at 11 02 08" src="https://user-images.githubusercontent.com/612155/39609224-a6a78b58-4f8a-11e8-98a2-f6ed0d81f85b.png">
<img width="73" alt="screen shot 2018-05-04 at 11 03 34" src="https://user-images.githubusercontent.com/612155/39609256-e8a6728a-4f8a-11e8-9921-91dd35d94589.png">
<img width="274" alt="screen shot 2018-05-04 at 11 06 25" src="https://user-images.githubusercontent.com/612155/39609292-3a5aeeee-4f8b-11e8-98ed-3d0405181985.png">

There's a thick black border in a button's pressed/active state:

<img width="93" alt="screen shot 2018-05-04 at 11 04 11" src="https://user-images.githubusercontent.com/612155/39609267-01b895d2-4f8b-11e8-93bd-502d3cd7a06c.png">

jasmussen commented:
Thanks for doing this. I'll see if I can get time today to push a few visual fixes directly to this branch to address Roberts notes.

jasmussen commented:
Pushed a fix to the alignment issue. Will look at some other issues with focus styles and admin theme colors in a minute. 

@afercia 

> when I try to make an IconButton have also some text, the position of the icon is OK if the button is a real button, but it's not OK when it's actually a link. Seems it happens in Chrome, Firefox is OK. Seems the text-indent: 4px; doesn't work for links in Chrome. See the Preview button I've made it an IconButton for testing purposes:

Is this specific to _this PR_ or was it also an issue before? The text-indent is definitely there to indent the text when there's an icon, and combined with flex it seems to have worked decently well. However if it doesn't work well in all situations or browsers, then I can take a look (as I added this initially). Feel free to create a new issue and assign me, and I'll look at it if it's not specific to this branch.

jasmussen commented:
@youknowriad The theme colors regressed in this, because the CSS classes changed and additional specificity is required. That means the red color that the midnight theme applies to buttons with `button-primary` no longer works. 

I _can_ fix that in our `_admin-schemes.scss` mixin, but this is going to get messy for a variety of reasons, so before I do that, is there any other way we can approach this, so the core WordPress styles can still theme this?  

Otherwise we're overriding buttons, _and_ overriding theme styles. We're already doing this in a few places, sure, but if this is the path forward, we might as well detach these buttons as much as possible from the core UI, and see how many `!important`'s we can remove in the process. 

One other challenge with overriding the theme styles is additional colors. In the admin schemes mixin I've tried to limit things to _two_ colors, the primary spot color and the secondary spot color. I've then used SCSS darken to create the additional shades necessary for the candy-colored "saving" indicator. 

For the primary button I'm going to need all sorts of shades of the spot-color, for the border, for the pressed state, for the shadow, even the text shadow. I can try and re-create these using darken, but that won't be 100% accurate. The alternative is to duplicate every is-primary style for every admin scheme there is — but then we're duplicating yet again. 

Is there any way we can keep using the core "button-primary" styles, instead of re-creating the `is-primary` style?

youknowriad commented:
> we might as well detach these buttons as much as possible from the core UI, and see how many !important's we can remove in the process.

I'm fine with that, that's exactly what this PR is doing, just leaving the same styles for now.

> I can fix that in our _admin-schemes.scss mixin,

I think we can make try to automatically generated this mixin in the future. Like browsing our stylesheets and automatically add additional styles when we detect the usage of theme colors (similar to the RTL plugin). I think with good webpack skills (cc @aduth 😄 ), we can achieve that and don't think about these admin styles anymore. The good thing about such approach is that the theme styles would be loaded by stylesheet and not globally in a single stylesheet.

> Is there any way we can keep using the core "button-primary" styles, instead of re-creating the is-primary style?

That defeats the purpose of this PR. We'll style rely on Core style in that case which needs specificity and break reusability.

jasmussen commented:
I'll keep thinking about this and take a look next week if I have bandwidth.

youknowriad committed:
Fix admin schemes by using theme colors

youknowriad commented:
Now that the post-css plugin is merged, I rebased this PR and updated the primary buttons styles to use theme colors. They don't match the current buttons perfectly but they're close enough. What do you think @jasmussen

Commit:
Fix issues with hover, active, focus states.

This fixes #6668, inspired by @nfmohit work in #6714 (props).

jasmussen commented:
I pushed some fixes, Riad. Can you take a look and see how it feels? Otherwise, feels like we're VERY close!

youknowriad commented:
Looks good to me :) Let's get some eyes and keep this for the next release.

jasmussen commented:
CC: @karmatosed let me know if you'd like some GIFs as the buttons change a little.

jasmussen commented:
Also, the focus styles changed, let's make sure these are kosher.

karmatosed reviewed (approved):
Looks great, thanks.

aduth commented on components/button/style.scss:
```diff
@@ -1,10 +1,143 @@
+
+	/* Buttons that look like links, for a cross of good semantics with the visual */
+	&.is-link {
+		margin: 0;
```
Regression: This selector is more specific than (well, equal specific but out of order from) and thus overrides the margin applied from `.components-color-palette_clear`:

https://github.com/WordPress/gutenberg/blob/94889af43baf5142c52f79190947492bc86f9567/components/color-palette/style.scss#L9

v4.9.2|#6562
---|---
![v4.9.2](https://user-images.githubusercontent.com/1779930/40931233-55e74722-67f8-11e8-99f8-86613cb74364.png)|![#6562](https://user-images.githubusercontent.com/1779930/40931242-619d5bb0-67f8-11e8-8a39-0bba2b9c8c03.png)

aduth replied:
Honestly, we shouldn't need margin here for the button. The color palette panel shouldn't be doing a negative margin. It should have a container and spread the options via `display: flex` & `align-items: space-between`
